compiler-core/src/type_/expression.rs

```diff
@@ -166,9 +166,46 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 arguments: args,
             } => self.infer_record_update(*constructor, spread, args, location),
 
-            UntypedExpr::NegateBool { location, value } => self.infer_negate_bool(location, *value),
+            UntypedExpr::NegateBool { location, value } => {
+                if let UntypedExpr::NegateBool {
+                    location: _,
+                    value: _,
+                } = &*value
+                {
+                    self.environment
+                        .warnings
+                        .emit(Warning::DoubleUnary { location });
+                }
+                self.infer_negate_bool(location, *value)
```

Comment:
Could you put this logic inside the infer functions please, and have it check after type checking so we don't warn for invalid code.